lib/jekyll/site.rb

```diff
@@ -326,15 +326,16 @@ def relative_permalinks_are_deprecated
     #
     # Returns an Array of Documents which should be written
     def docs_to_write
-      @docs_to_write ||= documents.select(&:write?)
+      documents.select(&:write?)
     end
 
     # Get all the documents
     #
     # Returns an Array of all Documents
     def documents
-      @documents ||= collections.reduce(Set.new) do |docs, (_, collection)|
-        docs + collection.docs + collection.files
+      collections.each_with_object(Set.new) do |(_, collection), set|
+        collection.docs.each  { |item| set << item }
```

Comment:
Let's use `set.merge(collection.docs).merge(collection.files)`. [According to the documentation](https://ruby-doc.org/stdlib-2.6.3/libdoc/set/rdoc/Set.html#method-i-merge), it mutates the set. I don't see a reason to do this second iteration manually when the Ruby core team can add optimizations for us in the future 😄